### PR TITLE
Support scipy special function with tuple output

### PR DESCRIPTION
**Pull request by RandomY-2**
## What do these changes do?

We can use ExecutableTuple and TensorTupleElementOp to support mars execution of scipy special functions with tuple returns. Specifically, the following syntaxes can be supported:

```python
from mars.tensor import special
import mars.tensor as mt
x = mt.linspace(-3, 3)

S, C = special.fresnel(x)
S.execute()
C.execute()
special.fresnel(x).execute()
(S + C).execute()
```

Also since we invoke scipy.special.fresnel during tensor executions, a cache is added in core.py so if we execute multiple elements of the same function during one execute() call, we don't need to do the scipy computation for each element. Later executes can just use the cached scipy outputs.

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

**Comment by wjsi** [comment]
Given your implementation, it is not hard to use one single operand for two outputs. You may take a look at `TensorSVD` for an example, where we use `output_limit` property to define the number of outputs:

https://github.com/mars-project/mars/blob/424cfb9d422a0918a089821f2710340ff94e101a/mars/tensor/linalg/svd.py#L44-L46

use `new_tensors(..., kws[kw1, kw2])` to create multiple output objects:

https://github.com/mars-project/mars/blob/424cfb9d422a0918a089821f2710340ff94e101a/mars/tensor/linalg/svd.py#L69-L77

 and use `ctx[out.key] = xxx` to store execute results.

https://github.com/mars-project/mars/blob/424cfb9d422a0918a089821f2710340ff94e101a/mars/tensor/linalg/svd.py#L150-L155

**Comment by RandomY-2** [comment]
> Given your implementation, it is not hard to use one single operand for two outputs. You may take a look at `TensorSVD` for an example, where we use `output_limit` property to define the number of outputs:
> 
> https://github.com/mars-project/mars/blob/424cfb9d422a0918a089821f2710340ff94e101a/mars/tensor/linalg/svd.py#L44-L46
> 
> use `new_tensors(..., kws[kw1, kw2])` to create multiple output objects:
> 
> https://github.com/mars-project/mars/blob/424cfb9d422a0918a089821f2710340ff94e101a/mars/tensor/linalg/svd.py#L69-L78
> 
> and use `ctx[out.key] = xxx` to store execute results.
> 
> https://github.com/mars-project/mars/blob/424cfb9d422a0918a089821f2710340ff94e101a/mars/tensor/linalg/svd.py#L150-L155

Thank you for your feedback! Will take a look.

File: mars/tensor/special/core.py [hunk 1]
```diff
@@ -112,3 +115,54 @@ def execute(cls, ctx, op):
...
+
+        with device(device_id):
+            with np.errstate(**op.err):
+                if op.is_gpu():
+                    ret = cls._execute_gpu(op, xp, inputs[0])
+                else:
+                    ret = cls._execute_cpu(op, xp, inputs[0])
+
+                for i in range(len(op.outputs)):
+                    ctx[op.outputs[i].key] = ret[i]
```
Comment (wjsi) [inline mars/tensor/special/core.py:168]: Loop over `op.outputs` itself, not indices.

```suggestion
                for output in op.outputs:
                    ctx[output] = ret[i]
```

File: mars/tensor/special/core.py [hunk 2]
```diff
@@ -112,3 +115,54 @@ def execute(cls, ctx, op):
...
+                }
+                for i in range(len(res))
+            ],
+        )
+
+        if out is None:
+            return ExecutableTuple(res_tensors)
+
+        for i in range(len(res_tensors)):
+            out[i].data = res_tensors[i].data
```
Comment (wjsi) [inline mars/tensor/special/core.py:151]: Ditto.

File: mars/tensor/special/core.py [hunk 3]
```diff
@@ -112,3 +115,54 @@ def execute(cls, ctx, op):
...
+        res = func(np.ones(x.shape, dtype=x.dtype))
+        res_tensors = self.new_tensors(
+            [x],
+            kws=[
+                {
+                    "side": f"{self._func_name}[{i}]",
+                    "dtype": res[i].dtype,
+                    "shape": res[i].shape,
+                }
+                for i in range(len(res))
```
Comment (wjsi) [inline mars/tensor/special/core.py:143]: Ditto.

File: mars/tensor/special/tests/test_special.py [hunk 4]
```diff
@@ -276,6 +279,44 @@ def test_erfcinv():
...
+    raw = np.random.rand(10, 8, 5)
+    t = tensor(raw, chunk_size=3)
+
+    r = fresnel(t)
+    expect = scipy_fresnel(raw)
+
+    assert isinstance(r, ExecutableTuple)
+    assert len(r) == 2
+
+    for i in range(len(r)):
```
Comment (wjsi) [inline mars/tensor/special/tests/test_special.py:292]: Ditto.

File: mars/tensor/special/tests/test_special.py [hunk 5]
```diff
@@ -276,6 +279,44 @@ def test_erfcinv():
...
+    out = ExecutableTuple([t, t])
+    r_out = fresnel(t, out=out)
+
+    assert isinstance(out, ExecutableTuple)
+    assert isinstance(r_out, ExecutableTuple)
+
+    assert len(out) == 2
+    assert len(r_out) == 2
+
+    for i in range(len(r_out)):
```
Comment (wjsi) [inline mars/tensor/special/tests/test_special.py:310]: Ditto.

File: mars/tensor/special/tests/test_special_execution.py [hunk 6]
```diff
@@ -299,3 +299,25 @@ def test_quintuple_execution(setup, func):
...
+
+    raw = np.random.rand(10, 8, 6)
+    a = tensor(raw, chunk_size=3)
+
+    r = mt_func(a)
+
+    result = r.execute().fetch()
+    expected = sp_func(raw)
+
+    for i in range(len(result)):
```
Comment (wjsi) [inline mars/tensor/special/tests/test_special_execution.py:322]: Ditto.

File: mars/tensor/special/err_fresnel.py [hunk 7]
```diff
@@ -55,6 +59,12 @@ class TensorErfcinv(TensorSpecialUnaryOp):
     _func_name = "erfcinv"
 
 
+@_register_special_op
+class TensorFresnel(TensorTupleOp):
+    _func_name = "fresnel"
+    _func_outputs = 2
```
Comment (wjsi) [inline mars/tensor/special/err_fresnel.py:65]: Rename to `_n_outputs` might be better.

**Review by wjsi — COMMENTED** [review]
Code structure LGTM. Do not use `for i in range(len(list_obj))`. Using `for i, obj in enumerate(obj_list)` when indices are needed or `for obj in obj_list` is more pythonic.

**Review by wjsi — APPROVED** [review]
LGTM

**Review by hekaisheng — APPROVED** [review]
LGTM

**Pull request merged**